### ISSUE TITLE
[Bug]: NSInvalidUnarchiveOperationException  *** -[NSKeyedUnarchiver decodeObjectForKey:]: cannot decode object of class (OSIndirectNotification) for key (NS.objects) because no class named "OSIndirectNotification" was found; the class needs to be defined in source code or linked in from a library (ensure the class is part of the correct target). If the class was renamed, use setClassName:forClass: to add a class translation mapping to NSKeyedUnarchiver

### ISSUE THREAD

**Issue by marcosnovaesq** [OneSignal/OneSignal-iOS-SDK#1098]
### What happened?

We updated our react-native-onesignal lib that used OneSignal-iOS-SDK version 2.13.0, which had this bug, according to [this issue](https://github.com/OneSignal/OneSignal-iOS-SDK/issues/680) and according to [this one](https://github.com/OneSignal/OneSignal-iOS-SDK/issues/833) it got solved and later versions. We updated react-native-onesignal to version 4.3.7 ( uses sdk version 3.10.1)and did the correct steps to make it work in ios. Aparrently it started happeningto some of our iOS users. Digging in the code we found that it was supossed to run a migration on some version, but aparrently on our end it didn't happen. Anyway our test users that had this happening reinstalled and began to work, but not a fix for common production user

### Steps to reproduce?

```Markdown
Could not reproduce in simulator and in production not all of our ios users are having this error
```

### What did you expect to happen?

Work ok for all ios users

### OneSignal iOS SDK version

3.10.1

### iOS version

15

### Specific iOS version

```Markdown
happens in mostly iOS 15.x versions
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by marcosnovaesq** [comment]
i'm having crash with a lot of users on production, can someone help me?

**Comment by marcosnovaesq** [comment]
@jkasten2 yes, when users reinstall the app it start working again as supposed to... Any info i could pass you to repro? We are with this bug for 2 weeks already

**Comment by marcosnovaesq** [comment]
according to method migrateToVersion_02_14_00_AndGreater in class OSMigrationController.m if the user had a cached sdk version lower than 21400 then it should migrate... Before the major update we did we had onesignal-ios-sdk on version 2130. Maye it's the cause, i just don't know why users aren't getting migrated...

**Comment by pedrosbmartins** [comment]
+1

I'm having this exact issue. Recently updated from 3.7.3 to 4.3.7 and some iOS users randomly crash with

> -[NSKeyedUnarchiver decodeObjectForKey:]: cannot decode object of class (OSIndirectNotification) for key (NS.objects) because no class named "OSIndirectNotification" was found

Asking for a complete reinstall solves the issue, but that's not a valid solution to the entire user base.

Unfortunately, we haven't been able to reproduce the issue on the simulator either.

**Comment by marcosnovaesq** [comment]
any updates on this?

**Comment by marcosnovaesq** [comment]
Updated: we released a version with the one signal completely removed from the app and the user stopped seeing the crash. We then released another version with it and started crashing again. So i don't know if you plan to fix this issue again or not, but for the moment we are going without one signal plugged in our app

**Comment by emawby** [comment]
@marcosnovaesq I apologize for the inconvenience we are actively working in reproducing and fixing this issue

**Comment by jkasten2** [comment]
@marcosnovaesq @pedrosbmartins After researching the issue I found a case where this issue could happen, if you updated the OneSignal SDK i updated on the main target but an older version is still being used with `OneSignalNotificationServiceExtension`.

Can you check your `Podfile` to ensure it has the following?:
```ruby
target 'OneSignalNotificationServiceExtension' do
  pod 'OneSignalXCFramework', '>= 3.4.3', '< 4.0'
end
```

Make sure to remove `pod 'OneSignal'` as well.

**Comment by marcosnovaesq** [comment]
@jkasten2 On my side, when we upgraded to another major, we did not change this line, at first. When I was fixing the bug, changing this line was one of my many failed attempts xD...  But we didn't change this line when we upgraded, only later. I only managed to fix when i released a version with react-native-onesignal 3.8.1

**Comment by jkasten2** [comment]
@marcosnovaesq Thanks for following up, however I am not sure I understand. Sounds like you ran into some issue upgrading that line in your `Podfile` but got it updated in the end?

Now that everything is updated you are no longer seeing this issue now?

**Comment by marcosnovaesq** [comment]
Sorry for the confusion, what happened was
-> updated one signal from 3.x.x to 4.x.x and that line was not changed
-> issue started happening in production
-> first attempt to solve it was to change this line as it was the only difference from the update guide of your documentation https://documentation.onesignal.com/docs/step-by-step-react-native-3x-to-400-upgrade-guide

-> issue still happening, crash still being caught by sentry

**Comment by jkasten2** [comment]
@marcosnovaesq I see, that is clear to me now!

My best guess what is happening now is that the migration was already run but the old OneSignal version in the NSE (NotificationServiceExtension) added some records with the old version. Now that the NSE is updated it isn't writing the old format, however now there are some stuck old records continuing to create this crash.

We will look into what can be done to address this state your app and possibly some others are stuck in due to this mix of OneSignal versions.

**Comment by marcosnovaesq** [comment]
thanks @jkasten2  i thought it weird too, looked at the code and the migration should have run, maybe it just didn't migrate correctly? Maybe a edge case where you migrate to buggy 3.x.x to major 4, i don't know. Are you planning on release something to address it? Thanks!!

**Comment by jkasten2** [comment]
@marcosnovaesq I believe your user got migrated correct on the first time the app ran to upgrade them. But the issue is the old NSE put some old record on the device and the migration won't run a 2nd time currently to fix those records. We are looking into addressing this edge case in the next release (3.11.3 or 3.12.0).

**Comment by marcosnovaesq** [comment]
Oh i see! it makes sense since that line was not updated together, it used old NSE... nice

**Comment by marcosnovaesq** [comment]
Hello @jkasten2 how is it going? any news? Thanks

**Comment by jkasten2** [comment]
@marcosnovaesq Sorry for the delay, we don't have an eta yet on this fix.

**Comment by angyy10** [comment]
Hi our existing iOS app user get this error as well, even updating app to latest OneSginal SDK version didn't help, when is this issue getting fixed?

**Comment by marcosnovaesq** [comment]
@angyy10 according to our interpretation we should only wait for them to release a fix for this... But no estimates so far, let me know if you fix this

**Comment by marcosnovaesq** [comment]
Any news?

**Comment by marcosnovaesq** [comment]
Hello, are you still looking into this?

**Comment by marcosnovaesq** [comment]
Hello, anyone looking into this can provide more information regarding the current situation?
